Introduce Marshall.LoadLibrary API

swaroop-sridhar opened the pull request.

jkotas commented on src/vm/marshalnative.h:
```diff
@@ -85,6 +85,16 @@ class MarshalNative
+    static INT_PTR QCALLTYPE LoadLibraryFromPath(LPCWSTR path, BOOL throwOnError);
+    static INT_PTR QCALLTYPE LoadLibraryByName(LPCWSTR name, QCall::AssemblyHandle callingAssembly, 
+                                                              DWORD dllImportSearchPathFlag, BOOL throwOnError);
+    static void QCALLTYPE FreeNativeLibrary(INT_PTR handle, BOOL throwOnError);
```
FreeNativeLibrary should not have throwOnError argument.

jkotas commented on src/vm/marshalnative.cpp:
```diff
@@ -923,6 +923,133 @@ FCIMPL1(int, MarshalNative::GetHRForException_WinRT, Object* eUNSAFE)
+    // PAL_GetProcAddressDirect() accepts wide-string procedure name, 
+    // in line with other public PAL API.
+
+    address = reinterpret_cast<INT_PTR>(PAL_GetProcAddressDirect((NATIVE_LIBRARY_HANDLE)handle, symbolName));
```
It would be better to `PAL_GetProcAddressDirect` to also take the entrypoint as LPCSTR. It would reduce amount of ifdefed out code.

jkotas commented on src/vm/marshalnative.cpp:
```diff
@@ -923,6 +923,133 @@ FCIMPL1(int, MarshalNative::GetHRForException_WinRT, Object* eUNSAFE)
+#ifndef FEATURE_PAL
+    // GetProcAddress only takes  LPCSTR for procedure name.
+    // Therefore, convert symbolName to a multi-byte-char encoding before calling GetProcAddress()
+    DWORD wideLen = (DWORD) wcslen(symbolName);
```
We have helpers for this (`MAKE_UTF8PTR_FROMWIDE` or `SString`).

jkotas commented on src/vm/marshalnative.cpp:
```diff
@@ -923,6 +923,133 @@ FCIMPL1(int, MarshalNative::GetHRForException_WinRT, Object* eUNSAFE)
+    if(throwOnError && (address == NULL))
+    {
+#ifdef FEATURE_PAL
+        COMPlusThrow(kEntryPointNotFoundException, IDS_EE_NDIRECT_GETPROCADDRESS_UNIX, W("GetLibraryExport_handle"), symbolName);
```
What does the `GetLibraryExport_handle` refer to?

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6574,7 +6611,7 @@ HINSTANCE NDirect::LoadLibraryModule(NDirectMethodDesc * pMD, LoadLibErrorTracke
             hmod = PAL_RegisterLibraryDirect(hmod, wszLibName);
```
(unrelated to the current change) I am wondering whether we need the `PAL_RegisterLibraryDirect` at all after your refactoring. Looks like we may be able to switch the rest of this to use the `NATIVE_LIBRARY_HANDLE` directly as well.

swaroop-sridhar replied:
@jkotas  and I talked about this. @jkotas said that we can do away with the call to DllMain on Unix systems, thus making the PAL abstraction on native handles unnecessary. This will be addressed in a separate change. CC: @janvorli

swaroop-sridhar replied:
https://github.com/dotnet/coreclr/issues/21009 Tracks this.

swaroop-sridhar commented:
@Jkotas, I've addressed your feedback in the new commit, and added test cases in `Interop/MarshalAPI/NativeLibrary`

jkotas commented on src/System.Private.CoreLib/Resources/Strings.resx:
```diff
@@ -382,6 +382,9 @@
   <data name="Arg_InvalidANSIString" xml:space="preserve">
     <value>The ANSI string passed in could not be converted from the default ANSI code page to Unicode.</value>
   </data>
+  <data name="Arg_InvalidWideSting" xml:space="preserve">
```
Typo: Sting

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6132,22 +6136,107 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+
+    if(lpstr == NULL)
+    {
+        if(throwOnError)
```
The convention for `TryXXX` APIs is to throw for invalid input. They should skip throwing exception only if the operation itself fails.

For example, `Dictionary.TryGetValue(null, out ...)` is throws `ArgumentNullException`.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6132,22 +6136,107 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+
+    if(handle == NULL) 
+    {
+        if(throwOnError)
```
Nit: Add space between `if` and `(`

(Multiple places)

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1755,5 +1755,117 @@ public static void ZeroFreeGlobalAllocUnicode(IntPtr s)
+        /// This method is a wrapper around OS loader, using "default" flags.
+        /// </summary>
+        /// <param name="libraryName">The name of the native library to be loaded</param>
+        /// <returns>The NativeLibrary object for the loaded library</returns>  
```
This doesn't return the "NativeLibrary object" - please update.

AaronRobinsonMSFT replied:
Throughout.

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1755,5 +1755,117 @@ public static void ZeroFreeGlobalAllocUnicode(IntPtr s)
+        /// <exception cref="System.BadImageFormatException">Thrown if the library is not valid.</exception>
+        public static IntPtr LoadLibrary(string libraryPath)
+        {
+            return LoadLibraryFromPath(libraryPath, true);
```
Just for the sake of clarity, can you use named parameters for boolean types if possible?

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1755,5 +1755,117 @@ public static void ZeroFreeGlobalAllocUnicode(IntPtr s)
+        /// <returns>The NativeLibrary object for the loaded library</returns>  
+        /// <exception cref="System.DllNotFoundException ">Thrown if the library can't be found.</exception>
+        /// <exception cref="System.BadImageFormatException">Thrown if the library is not valid.</exception>        
+        public static IntPtr LoadLibrary(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
```
This API is weird.  Why make an enum value null? Can we just use the default value expected? In this case `DllImportSearchPath.LegacyBehavior`?

jkotas replied:
Wasn't the `DllImportSearchPath?` meant to distinguish between the case where you actually want the legacy behavior and the case where you want the flags to be fetched from the assembly?

As the code is written, we do not really need the Nullable there. Folks can just pass  `default(DllImportSearchPath)` to get the default.

cc @GrabYourPitchforks

morganbr replied:
Legacy shouldn't be the default since it's insecure. Null meaning check assembly attributes could be ok, but if the assembly doesn't have attributes, I'm not sure what we do. What's the problem with just requiring the user to choose?

swaroop-sridhar replied:
Sure, I'll change the semantics of DllImportSearchPath being null to mean that the DllImportSearchPathFlags should be read from the module (and not LegacyBehavior).

This fits well with the existing code, because current implementation is
- Look up the `DefaultDllImportSearchPathAttribute` on the `PInvoke` is available use-it. With the `Marshall.LoadLibrary()` API this will mean if `DllImportSearchPathFlag` parameter is specified use it.
- If not, Look up the `DefaultDllImportSearchPathAttribute` on the module, and use it if that's available.
- If not, use default behavior specified [here](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.defaultdllimportsearchpathsattribute?redirectedfrom=MSDN&view=netframework-4.7.2)

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1755,5 +1755,117 @@ public static void ZeroFreeGlobalAllocUnicode(IntPtr s)
+        /// <exception cref="System.BadImageFormatException">Thrown if the library is not valid.</exception>        
+        public static IntPtr LoadLibrary(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
+        {
+            uint searchPathFlag = searchPath.HasValue ? (uint)searchPath.Value : 0;
```
Instead of 0 use the proper enum value for clarity.

AaronRobinsonMSFT replied:
Throughout.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6132,22 +6136,107 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+    BOOL retVal = PAL_FreeLibraryDirect(handle);
+#endif // !FEATURE_PAL
+
+    if (retVal == 0)
```
I thought we decided to not throw if we failed to free the library? Same for the handle being null? (i.e. Free should never throw). I could be mis-remembering.

swaroop-sridhar replied:
@AaronRobinsonMSFT  and I talked about this. FreeLibrary will throw if the close operation fails (following the principle that dlclose and FreeLibrary() return error codes).
However, it won't throw on null input, to avoid further null propagation (ex: in finally blocks).

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6132,22 +6136,107 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+{
+    STANDARD_VM_CONTRACT;
+
+    INT_PTR address = NULL;
```
Move this much closer to usage.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6507,6 +6573,40 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleBySearch(NDirectMethodDesc * pMD
+// Search for the library and variants of its name in probing directories.
+// This method returns native system libray handle not registered with the PAL
+// static
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleBySearch(NDirectMethodDesc * pMD, LoadLibErrorTracker * pErrorTracker, const wchar_t* wszLibName)
```
I was under the impression that `wchar_t` should not be used in the code base and we should always use the `WCHAR` type?

jkotas replied:
`wchar_t` is not strictly required - we `#define` it in the PAL to be `WCHAR`.

In this case, it would be better to use `PCWSTR` to make it clear that it is a zero terminated string.

AaronRobinsonMSFT commented on tests/src/Common/CoreCLRTestLibrary/Utilities.cs:
```diff
@@ -57,6 +57,8 @@ public static bool Verbose
 
+        public static bool IsMacOSX => (Environment.OSVersion.Platform == PlatformID.MacOSX);
+        public static bool IsUnix => (Environment.OSVersion.Platform == PlatformID.Unix);
         public static bool IsWindows => (Path.DirectorySeparatorChar == '\\');
```
Could you update the Windows check so it is the same type of check?

swaroop-sridhar replied:
@AaronRobinsonMSFT  I changed the OS checks to use RuntimeInformation.IsOSPlatform() since Environment.OSVersion.Platform is [not reliable](https://docs.microsoft.com/en-us/dotnet/api/system.environment.osversion?view=netframework-4.7.2). It didn't work correctly for Mac.

AaronRobinsonMSFT commented on tests/src/Interop/MarshalAPI/NativeLibrary/NativeLibrary.cpp:
```diff
@@ -0,0 +1,20 @@
+#include <stdio.h>
+#include <xplatform.h>
+
+extern "C" DLL_EXPORT int RetValue = 100;
```
Why is this for?

AaronRobinsonMSFT commented on tests/src/Interop/MarshalAPI/NativeLibrary/NativeLibrary.cpp:
```diff
@@ -0,0 +1,20 @@
+
+extern "C" DLL_EXPORT int RetValue = 100;
+
+extern "C" DLL_EXPORT int __cdecl NativeSum(int a, int b)
```
Why cdecl? Unless there is a compelling reason, we should not need to define the calling convention unless testing something related to it.

AaronRobinsonMSFT replied:
Feel free to mark it as `STDCALLMETHOD` if you really want to be specific.

AaronRobinsonMSFT commented on tests/src/Interop/MarshalAPI/NativeLibrary/NativeLibraryTests.cs:
```diff
@@ -0,0 +1,396 @@
+        success &= EXPECT(TryGetLibraryExport(handle, "NativeSum"));
+
+        // Call with null handle
+        success &= EXPECT(GetLibraryExport((IntPtr) 0, "NativeSum"), TestResult.ArgumentNull);
```
`IntPtr.Zero`

AaronRobinsonMSFT commented on tests/src/Interop/MarshalAPI/NativeLibrary/NativeLibraryTests.cs:
```diff
@@ -0,0 +1,396 @@
+        success &= EXPECT(FreeLibrary(handle), TestResult.InvalidOperation);
+
+        // Null Free
+        success &= EXPECT(FreeLibrary((IntPtr) null), TestResult.ArgumentNull);
```
`IntPtr.Zero`

AaronRobinsonMSFT commented on tests/src/Interop/MarshalAPI/NativeLibrary/NativeLibraryTests.cs:
```diff
@@ -0,0 +1,396 @@
+            return TestResult.GenericException;
+        }
+
+        if (handle == null)
```
`handle == IntPtr.Zero`

morganbr commented on src/vm/dllimport.cpp:
```diff
@@ -6132,22 +6136,107 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
 
     LoadLibErrorTracker errorTracker;
     const NATIVE_LIBRARY_HANDLE hmod =
         LocalLoadLibraryHelper(libraryPath, GetLoadWithAlteredSearchPathFlag(), &errorTracker);
```
I dont' think this matches the semantic of the callers. Altered search path looks in a bunch of different places and isn't secure.

swaroop-sridhar replied:
The [only difference](https://docs.microsoft.com/en-us/windows/desktop/Dlls/dynamic-link-library-search-order) caused by the LOAD_WITH_ALTERED_SEARCH_PATH is: "The standard search begins in the calling application's directory, and the alternate search begins in the directory of the executable module that LoadLibraryEx is loading."

I kept the LOAD_WITH_ALTERED_SEARCH_PATH in this case because it was already used by the AssemblyLoadContext.LoadUnmanagedDll() implementation. I think that these two APIs should at least agree on what the default flags are (hence they use the same method for loading).

We can pick the default without LOAD_WITH_ALTERED_SEARCH_PATH for both Marshall.LoadLibrary() and AssemblyLoadContext.LoadUnmanagedDll() if we think that is better. I'm not really sure if one is clearly better. But I think we should make this change separately since it modifies existing behavior.

Note that when we try to load from NATIVE_DLL_SEARCH_DIRECTORIES (given  by the host), we also LOAD_WITH_ALTERED_SEARCH_PATH https://github.com/dotnet/coreclr/blob/master/src/vm/dllimport.cpp#L6416

swaroop-sridhar replied:
When DllImportSearchPathFlags are absent (as in this case), the documented behavior is to use LOAD_WITH_ALTERED_SEARCH_PATH.

"The common language runtime handles the call to the LoadLibraryEx function according to the following algorithm:

If the attribute is applied to an individual platform invoke, use the values specified by that instance of the attribute.

Otherwise, if the attribute is applied to the assembly that contains the platform invoke, use the values specified by that instance of the attribute.

Otherwise, search the assembly directory and then call the LoadLibraryEx function with the LOAD_WITH_ALTERED_SEARCH_PATH flag.
"

morganbr commented on tests/src/Interop/MarshalAPI/NativeLibrary/NativeLibraryTests.cs:
```diff
@@ -0,0 +1,396 @@
+    static string CurrentTest;
+    static bool Verbose = false;
+ 
+    public static int Main()
```
Please add negative tests that ensure we don't load libraries from places that we shouldn't. E.g. that we don't load from the application or working directory if given a full path and that we're correctly piping DllImportSearchPath flags

swaroop-sridhar commented:
@Jkotas @AaronRobinsonMSFT @morganbr I've addressed the feedback in this checkin https://github.com/dotnet/coreclr/pull/20871/commits/d7c76828358bea7a741a41036900a840a92030e5.  [In a couple of cases, no changes were made, but I've left comments].

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6132,23 +6134,154 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+        BOOL attributeFound = GetDefaultDllImportSearchPathsAttributeValue(callingAssembly->GetManifestImport(),
+                                                                           TokenFromRid(1, mdtAssembly),
+                                                                           &dllImportSearchFlag);
+        if(attributeFound)
```
Nit: Add space between `if` and `(`.
(multiple places)

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6132,23 +6134,154 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+    // Otherwise checks if the assembly has the DefaultDllImportSearchPathsAttribute attribute. If so, use that value.
+    if (!hasDllImportSearchFlag)
+    {
+        BOOL attributeFound = GetDefaultDllImportSearchPathsAttributeValue(callingAssembly->GetManifestImport(),
```
You should use the cached value on Module.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6132,23 +6134,154 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+    if (symbolName == NULL)
+        COMPlusThrowArgumentNull(W("symbolName"), W("ArgumentNull_String"));
+
+    MAKE_UTF8PTR_FROMWIDE_NOTHROW(lpstr, symbolName);
```
Why not use `MAKE_UTF8PTR_FROMWIDE` that will throw the exception for you?

swaroop-sridhar commented:
@jkotas  thanks, I've fixed these issues.

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1760,5 +1760,133 @@ public static void ZeroFreeGlobalAllocUnicode(IntPtr s)
+        public static IntPtr LoadLibrary(string libraryPath)
+        {
+            bool throwOnError = true;
+            return LoadLibraryFromPath(libraryPath, throwOnError);
```
The normal way to write this is: `return LoadLibraryFromPath(libraryPath, throwOnError: true);`

You do not need to cache it in a local.

jkotas replied:
Multiple places.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6132,23 +6134,152 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
 {
     STANDARD_VM_CONTRACT;
 
+    if (libraryPath == NULL)
```
The usual way to do argument validation to do it as first thing in the publicly exposed method (e.g. either the C# method or QCall). Not bury it in the callers.

swaroop-sridhar replied:
@jkotas  I've made these changes in https://github.com/dotnet/coreclr/pull/21148
